### PR TITLE
Board Editor / Menu Issues

### LINKED ISSUE #1628: Mouse Wheel doesn't change Terrain Settings.
The tool tip says I can wheel to change terrain types, While the image in the bottom right changes. Everything in the Terrain Settings box stays the same. 

Effectively you can say you're using heavy woods. Something intermittently causes it to work and other times it defaults back to the base terrain.

image

### Environment
What version of MegaMek does your issue apply to?  47.4
What operating system are you using?  Windows

### Description
Describe the issue and provide screenshots ingame if able.
Provide the rulebook, page number, and text if you feel like this is a rules issue.
And provide the steps you have taken to reproduce the issue in the saved game if applicable.
(Delete the above lines)

### Files
Provide a saved game of when the issue was found.  If possible try to save the game during the phase before the issue occurs so the 
developers can step through the code and find exactly what it is that is breaking in the code.
Provide the megameklog.txt file for any NPE's or errors that MegaMek might be reporting.
Do not paste the entire logfile text into this ticket. Just attach the actual megameklog.txt file
And provide your playername so that they can load the saved game.
(Delete the above lines)

### LINKED ISSUE #1857: 47.5 Map Editor - UI Improvement - suggest 'grey out' Exits value when checkbox unchecked.
v0.47.5

With the new UI, the 'Exits' value is permanently shown in the value box with the up/down buttons, even when the 'Exits' checkbox isn't active. This might be confusing to users. Is there a way to set the text colour in that box to a shade of grey when the 'Exits' checkbox is unchecked, and to prevent the value being changed?

### LINKED ISSUE #2181: RFE Map Editor - add "grass" as a positive theme selection for hexes
In map editors before 47.9, it was possible to select 'grass' as a theme for hexes. Now that is the default, it is no longer on the drop-down, and the 'null' option has to be seleted. However this does not encode the hex as being 'grass' themed, merely as 'null'-themed..
In previous editors, the default option was 'desert', which it was also impossible to positively select - instead the 'null' option had to be selected.
This change means that maps that were designed to appear thematically arid will now appear grassy. This will also mean that maps designed using the latest editor will look considerably different if played on the current stable release, as the 'null' hexes will appear to be grass in the new editor, and desert in the stable.

Please add "grass" as a positive option to the map editor theme drop-down, so that it is possible to select this as an option, rather than relying on the handling of the 'null' option.

### LINKED ISSUE #2264: [RFE] Map Editor - Keybind CTRL+Z to the undo button
Would it be possible to keybind the popular `CTRL+Z` key combination to the 'Undo' function on the map editor, now that we have an 'Undo' button?

### LINKED ISSUE #2290: [RFE] Map Editor - Advanced Level - Raise/lower entire map by x
This might be a useful tool for those wanting to adjust the 'base level' of a map without redrawing all of the terrain that they have already added.

This should hopefully be a simple case of iterating through every hex row in the .board file, increasing/decreasing the 'level' value by a selected value 'x'.

(Slightly niche request, I know! Maybe it could be put on the 'Resize map' dialogue, so as not to clutter the main GUI area.)

### LINKED ISSUE #2390: RFE - Map Editor "Change Tileset Theme" option to change 'Hex Settings' theme at same time
At present, while the "Change Tileset Theme" option in the Map Editor applies a change to the theme of each tile in the board, it does not change the 'theme' for the 'Hex Settings' (i.e. for the 'terrain brush') to match the new selected theme. Not a huge issue, but might be a minor usability tweak?

### LINKED ISSUE #2944: 48.0 Tooltip on File>Board>Save As Image... incorrect
Megamek 0.48.0 Standalone
Windows 10
Java version 11.0.8

As per issue title, the mouseover tooltip on **File > Board > Save As Image...** appears to be the tooltip for **File > Board > Save As Image with Units...**, as it states "Save the current board as a PNG image, with all units".

There is currently no tooltop for **Save As Image with Units**.

See illustration below ...
image

### PR DESCRIPTION
**Pull request by SJuliez**
Note: The issues correspond to individual commits as much as I could manage.

- Issue 1857: The exits value in the board editor is now grayed out when the exits checkbox is unselected and the exits value is thus inactive
- Issue 1628: Using the quick buttons now also selects their terrain in the terrain list when possible thus also setting up this terrain in the details text fields
- Issue 2264: Adds a menu entry for Undo and binds Ctrl-Z to it as widely used (at least on Windows)
- Issue 2944: Corrects the Save as image menu tooltip
- Issue 2390: When changing the board theme, the new theme is also set in the hex details for further painting
- Board Editor: Mouse wheeling a quick button when its terrain is not present no longer throws an NPE
- Issue 2181: Adds "grass" as a positive theme selection. Note that I just added a dummy entry to the Saxarba tilesets to make "grass" appear. When used it won't find any actual tiles for grass and will fall back to the default grass tiles.
- Moves out the file->board menu to its own Board menu
- Issue 2290: Adds a menu entry to raise/lower the entire board by a user-defined delta
- Adds a menu entry to flood the board up to a user-defined level
- Adds a menu entry to flatten the board, removing all levels
- Adds a menu entry to remove water, buildings, forests or roads from the board
- Adds a menu entry to clear the entire board 

SimonLandmine's excellent desert canyon after some rain (flood to level 3, change theme to grass, lower level by 6)
image

Resolves #1857
Resolves #1628
Resolves #2264 
Resolves #2390 
Resolves #2290 
Resolves #2181 
Resolves #2944

### CODE REVIEW

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java [hunk 1]
```diff
@@ -180,6 +179,35 @@ public CommonMenuBar() {
...
+        boardSaveAs = createMenuItem(menu, getString("CommonMenuBar.fileBoardSaveAs"), BOARD_SAVE_AS);
+        menu.addSeparator();
+        boardSaveAsImage = createMenuItem(menu, getString("CommonMenuBar.fileBoardSaveAsImage"), BOARD_SAVE_AS_IMAGE);
+        boardSaveAsImage.setToolTipText(getString("CommonMenuBar.fileBoardSaveAsImage.tooltip"));
+        boardSaveAsImageUnits = createMenuItem(menu, getString("CommonMenuBar.fileBoardSaveAsImageUnits"), BOARD_SAVE_AS_IMAGE_UNITS);
+        boardSaveAsImageUnits.setToolTipText(getString("CommonMenuBar.fileBoardSaveAsImageUnits.tooltip"));
+        menu.addSeparator();
+        boardUndo = createMenuItem(menu, getString("CommonMenuBar.boardUndo"), BOARD_UNDO); 
+        boardUndo.setAccelerator(KeyStroke.getKeyStroke(KeyEvent.VK_Z, InputEvent.CTRL_DOWN_MASK));
+        boardRedo = createMenuItem(menu, getString("CommonMenuBar.boardRedo"), BOARD_REDO);
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/CommonMenuBar.java:197]: Can we do a control y for redo?

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 2]
```diff
@@ -1995,12 +2001,174 @@ public void actionPerformed(ActionEvent ae) {
...
+            }
+            hasChanges = !canReturnToSaved | (undoStack.size() != savedUndoStackSize);
+        }
+    }
+    
+    /** Asks for a level delta and changes the level of all the board's hexes by that delta. */
+    private void boardChangeLevel() {
+        var dlg = new LevelChangeDialog(frame);
+        dlg.setVisible(true);
+        if ((dlg.getResult() != DialogResult.CONFIRMED) || (dlg.getLevelChange() == 0)) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:2110]:
```suggestion
        if (!dlg.getResult().isConfirmed() || (dlg.getLevelChange() == 0)) {
```

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 3]
```diff
@@ -1995,12 +2001,174 @@ public void actionPerformed(ActionEvent ae) {
...
+            redoStack.clear();
+            buttonRedo.setEnabled(false);
+            // When Undo (without Redo) has been used after saving
+            // and the user draws on the board, then it can
+            // no longer know if it's been returned to the saved state
+            // and it will always be treated as changed.
+            if (savedUndoStackSize > undoStack.size()) {
+                canReturnToSaved = false;
+            }
+            hasChanges = !canReturnToSaved | (undoStack.size() != savedUndoStackSize);
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:2102]:
```suggestion
            hasChanges = !canReturnToSaved || (undoStack.size() != savedUndoStackSize);
```

File: megamek/src/megamek/client/ui/swing/BoardEditor.java [hunk 4]
```diff
@@ -1995,12 +2001,174 @@ public void actionPerformed(ActionEvent ae) {
...
         }
+        correctExits();
+        endCurrentUndoSet();
     }
     
+    /** Asks for flooding info and then floods the whole board with water up to a level. */
+    private void boardFlood() {
+        var dlg = new FloodDialog(frame);
+        dlg.setVisible(true);
+        if (dlg.getResult() != DialogResult.CONFIRMED) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/BoardEditor.java:2132]:
```suggestion
        if (!dlg.getResult().isConfirmed()) {
```

**Review by Windchild292 — COMMENTED** [review]
In general looks good, just a few comments

File: megamek/src/megamek/client/ui/swing/CommonMenuBar.java (hunk 1 above)
Comment (SJuliez) [reply megamek/src/megamek/client/ui/swing/CommonMenuBar.java:197]: I wanted to but there's a keycommandbind with Ctrl-Y for toggling the unit labels that eats that shortcut (I tried). We should maybe make a pass at the standard key settings at some point to make them more in line with computing after 1989. Then we can add the Ctrl-Y.

**Comment by SJuliez** [comment]
Ok apart from the keybind, saved for later, the rest is in as suggested.

**Review by Windchild292 — APPROVED** [review]

**Comment by SimonLandmine** [comment]
That's quite some rain! :-) Excellent work, thanks!

**Pull request merged**